### PR TITLE
CP-17648: Implement PVS_farm.[introduce|forget]

### PR DESCRIPTION
**Pull request by johnelse**

### CODE REVIEW

**Comment by johnelse** [comment]
https://issues.citrite.net/browse/CP-17648

File: ocaml/test/test_pvs_farm.ml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	let pvs_farm = Xapi_pvs_farm.introduce ~__context ~name in
+	assert_equal name (Db.PVS_farm.get_name ~__context ~self:pvs_farm);
+	assert_equal [] (Db.PVS_farm.get_cache_storage ~__context ~self:pvs_farm)
+
+let test_forget () =
+	let __context = make_test_database () in
+	let pvs_farm = Xapi_pvs_farm.introduce ~__context ~name in
+	let pvs_proxy =
+		make_pvs_proxy ~__context ~farm:pvs_farm ~currently_attached:true () in
+	(* Forget should fail if there's a proxy. *)
```
Comment (robhoes) [inline ocaml/test/test_pvs_farm.ml]: It should fail if there is a "running" proxy only. Let's check that it succeeds if there is a proxy that is not running.

File: ocaml/xapi/xapi_pvs_farm.ml [hunk 2]
```diff
@@ -20,10 +20,27 @@ let not_implemented x =
...
 let forget ~__context ~self =
-  not_implemented "PVS_farm.forget"
+	let open Db_filter_types in
+	(* Check there are no running proxies. *)
+	if Db.PVS_proxy.get_refs_where ~__context
+		~expr:(And
+			((Eq (Field "farm", Literal (Ref.string_of self))),
+			(Eq (Field "currently_attached", Literal "true"))))
+		<> []
+	then raise Api_errors.(Server_error
```
Comment (robhoes) [inline ocaml/xapi/xapi_pvs_farm.ml:38]: Please define a new, specific, API error for this, and the ones below.

File: ocaml/test/test_pvs_farm.ml (hunk 1 above)
Comment (johnelse) [inline ocaml/test/test_pvs_farm.ml]: Good idea :+1:

**Comment by robhoes** [comment]
Good apart from the two bits above.

File: ocaml/test/test_pvs_farm.ml [hunk 3]
```diff
@@ -15,15 +15,65 @@
...
+	let pvs_farm = Xapi_pvs_farm.introduce ~__context ~name in
+	Xapi_pvs_farm.forget ~__context ~self:pvs_farm;
+	assert_equal (Db.is_valid_ref __context pvs_farm) false
+
+let test_forget_stopped_proxy () =
+	let __context = make_test_database () in
+	let pvs_farm = Xapi_pvs_farm.introduce ~__context ~name in
+	let (_: API.ref_PVS_proxy) =
+		make_pvs_proxy ~__context ~farm:pvs_farm ~currently_attached:false () in
+	Xapi_pvs_farm.forget ~__context ~self:pvs_farm
```
Comment (robhoes) [inline ocaml/test/test_pvs_farm.ml]: No assert needed here, like in `test_forget_ok`?

Comment (johnelse) [inline ocaml/test/test_pvs_farm.ml]: Yeah could do that - I'll add it.

**Comment by robhoes** [comment]
:shipit:

**Pull request merged**